sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -87,6 +87,62 @@ case class MapKeys(child: Expression)
   override def prettyName: String = "map_keys"
 }
 
+@ExpressionDescription(
+  usage = """_FUNC_(a1, a2) - Returns a merged array matching N-th element of first
+  array with the N-th element of second.""",
+  examples = """
+    Examples
+      > SELECT _FUNC_(array(1, 2, 3), array(2, 3, 4));
+        [[1, 2], [2, 3], [3, 4]]
+  """,
+  since = "2.4.0")
+case class Zip(left: Expression, right: Expression)
+  extends BinaryExpression with ExpectsInputTypes {
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(ArrayType, ArrayType)
+
+  override def dataType: DataType = ArrayType(left.dataType.asInstanceOf[ArrayType].elementType)
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    nullSafeCodeGen(ctx, ev, (arr1, arr2) => {
+      val i = ctx.freshName("i")
+      s"""
+      for (int $i = 0; $i < $arr1.numElements(); $i ++) {
+        if ($arr1.isNullAt($i)) {
+          ${ev.isNull} = true;
+        } else {
+          ${ev.value}[$i] = ($arr1[$i], $arr2[$i]);
```

Comment:
Does this code `($arr1[$i], $arr2[$i])` work in Java?

Reply:
Probably not, it is just a snippet.

Reply:
@kiszk Do you have a suggestion on how I can assign the value, since it is a tuple/pair of generic elements?

Reply:
One high level question. Do we have performance or other advantages when we generate Java code?

Reply:
I tought that I should always generate java code lmao (I'm faily new to Spark codebase)
Thanks!

Reply:
I might be wrong. We have the similar discussion [here](https://github.com/apache/spark/pull/21061#discussion_r181399858).
If we need to generate Java code, how about writing runtime routine in scala and calling the routine from Java?